File: back/src/echoroo/schemas/foundation_models.py
"""Schemas for foundation model metadata and runs."""

from __future__ import annotations

import datetime
from enum import Enum
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field

from echoroo.schemas.base import BaseSchema
from echoroo.schemas.datasets import Dataset
from echoroo.schemas.tags import Tag
from echoroo.schemas.users import SimpleUser

__all__ = [
    "DatasetFoundationModelSummary",
    "FoundationModel",
    "FoundationModelRun",
    "FoundationModelRunCreate",
    "FoundationModelRunProgress",
    "FoundationModelRunSpecies",
    "FoundationModelRunStatus",
]


class FoundationModel(BaseSchema):
    """Available foundation model metadata."""

    id: int = Field(..., exclude=True)
    uuid: UUID
    slug: str
    display_name: str
    provider: str
    version: str
    description: str | None = None
    default_confidence_threshold: float = 0.1
    is_active: bool = True


class FoundationModelRunStatus(str, Enum):
    """Run status enumeration."""

    QUEUED = "queued"
    RUNNING = "running"
    POST_PROCESSING = "post_processing"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class FoundationModelRunSpecies(BaseSchema):
    """Aggregated species summary for a run."""

    id: int = Field(..., exclude=True)
    foundation_model_run_id: int = Field(..., exclude=True)
    gbif_taxon_id: str | None = None
    annotation_tag_id: int | None = Field(default=None, exclude=True)
    tag: Tag | None = None
    scientific_name: str
    common_name_ja: str | None = None
    detection_count: int = 0
    avg_confidence: float = 0.0


class FoundationModelRun(BaseSchema):
    """Run record schema."""

    id: int = Field(..., exclude=True)
    uuid: UUID
    foundation_model_id: int = Field(..., exclude=True)
    dataset_id: int = Field(..., exclude=True)
    requested_by_id: UUID | None = Field(default=None, exclude=True)
    foundation_model: FoundationModel | None = None
    dataset: Dataset | None = None
    requested_by: SimpleUser | None = None
    status: FoundationModelRunStatus = FoundationModelRunStatus.QUEUED
    confidence_threshold: float = 0.1
    scope: dict[str, Any] | None = None
    progress: float = 0.0
    total_recordings: int = 0
    processed_recordings: int = 0
    total_clips: int = 0
    total_detections: int = 0
    classification_csv_path: str | None = None
    embedding_store_key: str | None = None
    summary: dict[str, Any] | None = None
    error: dict[str, Any] | None = None
    started_on: datetime.datetime | None = None
    completed_on: datetime.datetime | None = None
    species: list[FoundationModelRunSpecies] | None = None


class DatasetFoundationModelSummary(BaseModel):
    """Summary of latest runs per model for a dataset."""

    foundation_model: FoundationModel
    latest_run: FoundationModelRun | None = None
    last_completed_run: FoundationModelRun | None = None


class FoundationModelRunCreate(BaseModel):
    """Payload to create a new run."""

    dataset_uuid: UUID
    foundation_model_slug: str
    confidence_threshold: float | None = Field(default=None, ge=0.0, le=1.0)
    scope: dict[str, Any] | None = None
    locale: str = Field(
        default="ja",
        max_length=16,
        description="Locale for species common names (e.g., 'en_us', 'ja')",
    )
    """Locale for species common names. Defaults to Japanese ('ja')."""


class FoundationModelRunProgress(BaseModel):
    """Schema for tracking foundation model run progress."""

    status: FoundationModelRunStatus
    """Current run status."""

    progress: float = 0.0
    """Progress percentage (0.0 to 1.0)."""

    total_recordings: int = 0
    """Total recordings to process."""

    processed_recordings: int = 0
    """Recordings processed so far."""

    total_clips: int = 0
    """Total clips analyzed."""

    total_detections: int = 0
    """Detections found so far."""

    recordings_per_second: float | None = None
    """Processing speed."""

    estimated_time_remaining_seconds: float | None = None
    """Estimated time remaining."""

    message: str | None = None
    """Human-readable status message."""

File: front/src/lib/schemas/foundation_models.ts
"use client";

import { z } from "zod";

import { ClipSchema } from "./clips";
import { DatasetSchema } from "./datasets";
import { TagSchema } from "./tags";
import { SimpleUserSchema } from "./users";

export const FoundationModelSchema = z.object({
  uuid: z.string().uuid(),
  slug: z.string(),
  display_name: z.string(),
  provider: z.string(),
  version: z.string(),
  description: z.string().nullable().optional(),
  default_confidence_threshold: z.number(),
  is_active: z.boolean(),
  created_on: z.coerce.date(),
});

export const FoundationModelRunStatusSchema = z.enum([
  "queued",
  "running",
  "post_processing",
  "completed",
  "failed",
  "cancelled",
]);

export const FoundationModelRunSpeciesSchema = z.object({
  gbif_taxon_id: z.string().nullable().optional(),
  annotation_tag_id: z.number().nullable().optional(),
  tag: TagSchema.nullish(),
  scientific_name: z.string(),
  vernacular_name: z.string().nullable().optional(),
  detection_count: z.number(),
  avg_confidence: z.number(),
  created_on: z.coerce.date(),
});

export const FoundationModelRunSchema = z.object({
  uuid: z.string().uuid(),
  foundation_model_id: z.number().optional(),
  dataset_id: z.number().optional(),
  requested_by_id: z.string().uuid().nullable().optional(),
  foundation_model: FoundationModelSchema.nullish(),
  dataset: DatasetSchema.nullish(),
  requested_by: SimpleUserSchema.nullish(),
  status: FoundationModelRunStatusSchema,
  confidence_threshold: z.number(),
  scope: z.record(z.any()).nullable().optional(),
  progress: z.number().optional(),
  total_recordings: z.number().optional(),
  processed_recordings: z.number().optional(),
  total_clips: z.number().optional(),
  total_detections: z.number().optional(),
  classification_csv_path: z.string().nullable().optional(),
  embedding_store_key: z.string().nullable().optional(),
  summary: z.record(z.any()).nullable().optional(),
  error: z.record(z.any()).nullable().optional(),
  started_on: z.coerce.date().nullable().optional(),
  completed_on: z.coerce.date().nullable().optional(),
  created_on: z.coerce.date(),
  species: z.array(FoundationModelRunSpeciesSchema).nullable().optional(),
});

export const DatasetFoundationModelSummarySchema = z.object({
  foundation_model: FoundationModelSchema,
  latest_run: FoundationModelRunSchema.nullish(),
  last_completed_run: FoundationModelRunSchema.nullish(),
});

export const FoundationModelRunCreateSchema = z.object({
  dataset_uuid: z.string().uuid(),
  foundation_model_slug: z.string(),
  confidence_threshold: z.number().min(0).max(1).optional(),
  scope: z.record(z.any()).optional(),
  locale: z.string().optional(),
  run_embeddings: z.boolean().optional().default(true),
  run_predictions: z.boolean().optional().default(true),
});

// ============================================================================
// Foundation Model Run Progress
// ============================================================================

export const FoundationModelRunProgressSchema = z.object({
  status: FoundationModelRunStatusSchema,
  progress: z.number(),
  total_recordings: z.number().int(),
  processed_recordings: z.number().int(),
  total_clips: z.number().int(),
  total_detections: z.number().int(),
  recordings_per_second: z.number().nullable(),
  estimated_time_remaining_seconds: z.number().nullable(),
  message: z.string().nullable(),
});

// ============================================================================
// Foundation Model Detection Review Status
// ============================================================================

export const FoundationModelDetectionReviewStatusSchema = z.enum([
  "unreviewed",
  "confirmed",
  "rejected",
  "uncertain",
]);

// ============================================================================
// Foundation Model Detection Result
// ============================================================================

export const FoundationModelDetectionSchema = z.object({
  uuid: z.string().uuid(),
  clip_id: z.number().int(),
  clip: z.lazy(() => ClipSchema),
  species_tag: TagSchema,
  confidence: z.number(),
  review_status: FoundationModelDetectionReviewStatusSchema,
  reviewed_on: z.coerce.date().nullable(),
  reviewed_by_id: z.string().uuid().nullable(),
  notes: z.string().nullable(),
  converted_to_annotation: z.boolean(),
  is_included: z.boolean().nullable(),
  occurrence_probability: z.number().nullable(),
});

// ============================================================================
// Foundation Model Detection Review
// ============================================================================

export const FoundationModelDetectionReviewSchema = z.object({
  uuid: z.string().uuid(),
  // Note: clip_prediction_id and species_detection_job_id are excluded in backend response
  status: FoundationModelDetectionReviewStatusSchema,
  reviewed_by_id: z.string().uuid().nullable(),
  reviewed_on: z.coerce.date().nullable(),
  notes: z.string().nullable(),
  converted_to_annotation: z.boolean(),
  clip_annotation_id: z.number().int().nullable(),
});

export const FoundationModelDetectionReviewUpdateSchema = z.object({
  status: FoundationModelDetectionReviewStatusSchema,
  notes: z.string().optional(),
});

// ============================================================================
// Species Summary
// ============================================================================

export const FoundationModelSpeciesSummarySchema = z.object({
  tag_id: z.number().int(),
  tag_value: z.string(),
  total_detections: z.number().int(),
  confirmed_count: z.number().int(),
  rejected_count: z.number().int(),
  uncertain_count: z.number().int(),
  unreviewed_count: z.number().int(),
  average_confidence: z.number().nullable(),
});

// ============================================================================
// Detection Summary
// ============================================================================

export const FoundationModelDetectionSummarySchema = z.object({
  total_detections: z.number().int(),
  unique_species: z.number().int(),
  species_summary: z.array(FoundationModelSpeciesSummarySchema),
  total_reviewed: z.number().int(),
  total_confirmed: z.number().int(),
  total_rejected: z.number().int(),
  total_uncertain: z.number().int(),
  total_unreviewed: z.number().int(),
  // Optional fields (not currently computed by backend)
  confidence_histogram: z.array(z.number().int()).optional(),
  detections_by_date: z.record(z.number().int()).optional(),
  detections_by_location: z.record(z.number().int()).optional(),
});

// ============================================================================
// Bulk Review Response
// ============================================================================

export const BulkReviewResponseSchema = z.object({
  reviewed_count: z.number().int(),
});

// ============================================================================
// Job Queue Status
// ============================================================================

export const JobQueueStatusSchema = z.object({
  pending: z.number().int(),
  running: z.number().int(),
  completed: z.number().int(),
  failed: z.number().int(),
});

// ============================================================================
// Convert to Annotation Project Response
// ============================================================================

export const ConvertToAnnotationProjectResponseSchema = z.object({
  annotation_project_uuid: z.string().uuid(),
  annotation_project_name: z.string(),
  total_tasks_created: z.number().int(),
  total_annotations_created: z.number().int(),
});
